public/apps/configuration/panels/role-edit/role-edit.tsx

```diff
@@ -0,0 +1,163 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import {
+  EuiFieldText,
+  EuiForm,
+  EuiLink,
+  EuiPageHeader,
+  EuiSpacer,
+  EuiText,
+  EuiTitle,
+} from '@elastic/eui';
+import React, { useEffect, useState } from 'react';
+import { AppDependencies } from '../../../types';
+import { CLUSTER_PERMISSIONS, INDEX_PERMISSIONS } from '../../constants';
+import { fetchActionGroups } from '../../utils/action-groups-utils';
+import { stringToComboBoxOption } from '../../utils/combo-box-utils';
+import { FormRow } from '../../utils/form-row';
+import { PanelWithHeader } from '../../utils/panel-with-header';
+import { getRoleDetail } from '../../utils/role-detail-utils';
+import { buildIndexPermissionState, IndexPermissionPanel } from './index-permission-panel';
+import { TenantPanel } from './tenant-panel';
+import { ComboBoxOptions, RoleIndexPermissionStateClass } from './types';
+import { ClusterPermissionPanel } from './cluster-permission-panel';
+
+interface RoleEditDeps extends AppDependencies {
+  action: 'create' | 'edit' | 'duplicate';
+  // For creation, sourceRoleName should be empty string.
+  // For editing, sourceRoleName should be the name of the role to edit.
+  // For duplication, sourceRoleName should be the name of the role to copy from.
+  sourceRoleName: string;
+}
+
+const TITLE_TEXT_DICT = {
+  create: 'Create Role',
+  edit: 'Edit Role',
+  duplicate: 'Duplicate Role',
+};
+
+export function RoleEdit(props: RoleEditDeps) {
+  const [roleName, setRoleName] = useState('');
+  const [roleClusterPermission, setRoleClusterPermission] = useState<ComboBoxOptions>([]);
+  const [roleIndexPermission, setRoleIndexPermission] = useState<RoleIndexPermissionStateClass[]>(
+    []
+  );
+
+  useEffect(() => {
+    const action = props.action;
+    if (action == 'edit' || action == 'duplicate') {
+      const fetchData = async () => {
+        try {
+          const roleData = await getRoleDetail(props.coreStart.http, props.sourceRoleName);
+          setRoleClusterPermission(roleData.cluster_permissions.map(stringToComboBoxOption));
+          setRoleIndexPermission(buildIndexPermissionState(roleData.index_permissions));
+
+          if (action == 'edit') {
+            setRoleName(props.sourceRoleName);
+          } else {
+            setRoleName(props.sourceRoleName + '_copy');
```

Comment:
Just notice that users enter a unique name in this popup for duplication.
https://aws-uxdr.invisionapp.com/share/2JVZS7DVBCD?redirHash=#/screens/405153388

Thus when they reach the create/edit page, we should just use whatever sourceRoleName is.

Reply:
Current implementation is interim solution. The sourceRoleName has to be original role name so we can copy data. 
For the given screenshot, it directly clone it w/o going to the editing page.